File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [0] * n
        for i in range(n):
            if i == 0: 
                prefix[i] = nums[i]
                continue
            prefix[i] = prefix[i - 1] * nums[i]
        
        postfix = [0] * n
        for i in range(n - 1, -1, -1):
            if i == (n - 1): 
                postfix[i] = nums[i]
                continue
            postfix[i] = postfix[i + 1] * nums[i]

        result = [0] * n
        result[0] = postfix[1]
        result[-1] = prefix[-2]
        for i in range(1, n - 1):
            result[i] = postfix[i + 1] * prefix[i - 1]
        return result

File: Data Structures & Algorithms/subsets/submission-2.cpp
class Solution {
public:
    void backtrack(vector<int>& nums, vector<bool>& is_used, int start_idx) {
        if (temp.size() == nums.size()) {
            result.push_back(temp);
            return;
        }

        for (int i = start_idx; i <= nums.size(); ++i) {
            if (i == nums.size()) {
                result.push_back(temp);
                continue;
            } 
            if (!is_used[i]) {
                temp.push_back(nums[i]);
                is_used[i] = true;
                backtrack(nums, is_used, i + 1);
                is_used[i] = false;
                temp.pop_back();
            }
        }
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<bool> is_used(nums.size(), false);
        backtrack(nums, is_used, 0);
        return result;
    }

    vector<int> temp;
    vector<vector<int>> result;
};
